addons/account_edi/models/account_edi.py

```diff
@@ -0,0 +1,178 @@
+
+from odoo import models, fields, api
+from odoo.exceptions import UserError
+from odoo.tools.pdf import OdooPdfFileReader, OdooPdfFileWriter
+
+from lxml import etree
+import base64
+import io
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class AccountEdi(models.Model):
+    _name = 'account.edi'
+    _description = 'EDI format'
+
+    name = fields.Char(index=True)
+    code = fields.Char()
+    export_invoice = fields.Selection(selection=[('embed_to_pdf', 'Embed in PDF'), ('separated_attachment', 'Attachment'), ('not_supported', 'Not Supported')], 
+                                      default='embed_to_pdf')
+
+    _sql_constraints = [
+        ('unique_code', 'unique (code)', 'This code already exists')
+    ]
+
+    def _export_invoice_to_attachment(self, invoice):
+        """ Create the file content representing the invoice.
+
+        :param invoice: the invoice to encode.
+        :returns: a dictionnary (values are compatible to create an ir.attachment)
+        * name : the name of the file
+        * datas : the content of the file,
+        * res_model : 'account.move',
+        * res_id: the id of invoice
+        * mimetype : the mimetype of the attachment
+        """
+        # TO OVERRIDE
+        self.ensure_one()
+
+        return False
+
+    def _embed_attachments_into_report(self, pdf_content, invoice):
+        """ Create the EDI of the invoice and embed it in the pdf_content.
+
+        :param pdf_content: the bytes representing the pdf to add the EDIs to.
+        :param invoice: the invoice to generate the EDI from.
+        :returns: the same pdf_content with the EDI of the invoice embed in it.
+        """
+        reader_buffer = io.BytesIO(pdf_content)
+        reader = OdooPdfFileReader(reader_buffer)
+        writer = OdooPdfFileWriter()
+        writer.cloneReaderDocumentRoot(reader)
+
+        for edi in self.filtered(lambda e: e.export_invoice == 'embed_to_pdf'):
+            vals = edi._export_invoice_to_attachment(invoice)
+            if vals:
+                writer.addAttachment(vals['name'], vals['datas'])
+
+        buffer = io.BytesIO()
+        writer.write(buffer)
+        pdf_content = buffer.getvalue()
+        reader_buffer.close()
+        buffer.close()
+        return pdf_content
+
+    def _create_ir_attachments(self, invoice):
+        """ Create ir.attachment for the EDIs from invoice.
+
+        :param invoice: the invoice to generate the EDI from.
+        :returns: the ids of the created attachments.
+        """
+        attachments = []
+        for edi in self.filtered(lambda e: e.export_invoice == 'separated_attachment'):
+            vals = edi._export_invoice_to_attachment(invoice)
+            if vals:
+                vals['datas'] = base64.encodebytes(vals['datas'])
+                attachment = self.env['ir.attachment'].create(vals)
+                attachments.append(attachment.id)
+
+        return attachments
```

Comment:
do this part in batch:
```
attachment_vals_list = []
for edi in self.filtered(lambda e: e.export_invoice == 'separated_attachment'):
            vals = edi._export_invoice_to_attachment(invoice)
            if vals:
                vals['datas'] = base64.encodebytes(vals['datas'])
                attachment_vals_list.append(vals)
return self.env['ir.attachment'].create(attachment_vals_list)
```
Is there a reason to return ids instead of the records?